doc/cascadia/Json-Utility-API.md

```diff
@@ -0,0 +1,257 @@
+
+```c++
+std::string one;
+std::optional<std::string> two;
+
+JsonUtils::GetValue(json, one);
+// one is populated or unchanged.
+
+JsonUtils::GetValue(json, two);
+// two is populated, nullopt or unchanged
+
+auto three = JsonUtils::GetValue<std::string>(json);
+// three is populated or zero-initialized
+
+auto four = JsonUtils::GetValue<std::optional<std::string>>(json);
+// four is populated or nullopt
+```
+
+## Key lookup (GetValueForKey)
+
+`GetValueForKey` follows the same rules as `GetValue`, but takes an additional key.
+It is assumed that the JSON value passed to GetValueForKey is of `object` type.
+
+```c++
+std::string one;
+std::optional<std::string> two;
+
+JsonUtils::GetValueForKey(json, "firstKey", one);
+// one is populated or unchanged.
+
+JsonUtils::GetValueForKey(json, "secondKey", two);
+// two is populated, nullopt or unchanged
+
+auto three = JsonUtils::GetValueForKey<std::string>(json, "thirdKey");
+// three is populated or zero-initialized
+
+auto four = JsonUtils::GetValueForKey<std::optional<std::string>>(json, "fourthKey");
+// four is populated or nullopt
+```
+
+## Rationale: Value-Returning Getters
+
+JsonUtils provides two types of `GetValue...`: value-returning and reference-filling.
+
+The reference-filling fixtures use type deduction so that a developer does not
+need to specify template parameters on every `GetValue` call. It excels at
+populating class members during deserialization.
+
+The value-returning fixtures, on the other hand, are very useful for partial
+deserialization and key detection when you do not need to deserialize an entire
+instance of a class or you need to reason about the presence of members.
+
+To provide a concrete example of the latter, consider:
+
+```c++
+if (const auto guid{ GetValueForKey<std::optional<GUID>>(json, "guid") }) 
+    // This condition is only true if there was a "guid" member in the provided JSON object.
+    // It can be accessed through *guid.
+}
+```
+
+If you are... | Use
+--------------|-----
+Deserializing | `GetValue(..., storage)`
+Interrogating | `storage = GetValue<T>(...)`
+
+## Converting User-Defined Types
+
+All conversions are done using specializations of
+`JsonUtils::ConversionTrait<T>`.  To implement a converter for a user-defined
+type, you must implement a specialization of `JsonUtils::ConversionTrait<T>`.
+
+Every specialization over `T` must implement `static T FromJson(const Json::Value&)`
+and `static bool CanConvert(const Json::Value&)`.
+
+```c++
+struct MyCustomType { int val; };
+
+template<>
+struct ConversionTrait<MyCustomType>
+{
+    // This trait converts a string of the format "[0-9]" to a value of type MyCustomType.
+
+    static MyCustomType FromJson(const Json::Value& json)
+    {
+        return MyCustomType{ json.asString()[0] - '0' };
+    }
+
+    static bool CanConvert(const Json::Value& json)
+    {
+        return json.isString();
+    }

```

Comment:
So for settings that accept multiple types (i.e. `closeOnExit` accepts bool and string), this is where that gets defined?

Reply:
With regards to adopting this, would we have to create a converter for each of our settings?

The enum ones would use the `JSON_ENUM_MAPPER` below.

Reply:
We will need to create a converter for each setting that requires _special handling_. I've done all this work, but yeah. It's a big one.

The straight enum ones will be `JSON_ENUM_MAPPER`s, the custom enums will be (some hybrid)

Reply:
The idea is that we _generically_ specify how a specific _type_ (NOT A SPECIFIC OPTION (!)) is converted to/from json to increase standardization across the codebase

Any time we need to serialize or deserialize, one of these type-specific serializers will be used automatically, seamlessly.